### PR TITLE
[MD-1066] Container chain sending tokens to Ethereum

### PR DESCRIPTION
**Pull request by evgeny-s**
**Description**:
This PR provides a functionality for container chains to send tokens to Ethereum via the Relay chain

**NOTE**:
Should be merged after https://github.com/moondance-labs/tanssi/pull/1054

**The plan:**
- [x] Implement XCM Router on the container chain side
- [x] Cover container chains with dev tests
- [x] Implement/Re-use XCM Message Exporter on the relay chain side
- [x] E2E test
- [x] Documentation
- [x] Cover relay chain with dev tests
- [x] Cover corner cases with tests
- [x] Add more tests for potential vulnerabilities exposing

### CODE REVIEW

File: chains/orchestrator-relays/runtime/dancelight/src/xcm_config.rs [hunk 1]
```diff
@@ -198,25 +203,68 @@ impl Contains<Location> for LocalPlurality {
...
 }
 
+pub struct ParentOrParentsPlurality;
+impl Contains<Location> for ParentOrParentsPlurality {
+    fn contains(location: &Location) -> bool {
+        matches!(location.unpack(), (1, []) | (1, [Plurality { .. }]))
+    }
+}
+
+pub struct MockedAllowBarrier;
```
Comment (evgeny-s) [inline chains/orchestrator-relays/runtime/dancelight/src/xcm_config.rs:213]: This should be replaced

File: chains/container-chains/runtime-templates/frontier/src/lib.rs [hunk 2]
```diff
@@ -307,6 +309,23 @@ impl WeightToFeePolynomial for WeightToFee {
     }
 }
 
+parameter_types! {
+        /// Network and location for the Ethereum chain. On Starlight, the Ethereum chain bridged
+        /// to is the Ethereum mainnet, with chain ID 1.
+        /// <https://chainlist.org/chain/1>
+        /// <https://ethereum.org/en/developers/docs/apis/json-rpc/#net_version>
+        pub EthereumNetwork: NetworkId = NetworkId::Ethereum { chain_id: 11155111 };
```
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/container-chains/runtime-templates/frontier/src/lib.rs:317]: Or fetch it via the XCM message

File: chains/orchestrator-relays/runtime/dancelight/src/weights/xcm/pallet_xcm_benchmarks_generic.rs [hunk 3]
```diff
@@ -327,6 +327,30 @@ impl<T: frame_system::Config> WeightInfo<T> {
 		// Minimum execution time: 1_199_000 picoseconds.
 		Weight::from_parts(1_243_000, 0)
 	}
+	// Storage: `ParachainInfo::ParachainId` (r:1 w:0)
```
Comment (evgeny-s) [inline chains/orchestrator-relays/runtime/dancelight/src/weights/xcm/pallet_xcm_benchmarks_generic.rs:330]: Took it from BridgeHub for now, we will need to re-generate it on our side.

File: chains/container-chains/runtime-templates/frontier/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmpolaczyk) [reply chains/container-chains/runtime-templates/frontier/src/lib.rs:317]: Maybe having a testnet/mainnet flag could be useful for other things. But it adds an extra step to the deploy process... How can we make it automatic?

[the remark this answers is not in the document]
Comment (Agusrodri) [reply chains/container-chains/runtime-templates/frontier/src/lib.rs:317]: Just an idea: we could also have this `EthereumNetwork` as a param of `pallet_parameters`, and that way we would be able to modify it accordingly. This wouldn't be automatic tho, we can set a default value (e.g the testnet one) but if we want to change it in a mainnet environment, it should be done by calling the extrinsic.

File: chains/orchestrator-relays/runtime/dancelight/src/xcm_config.rs [hunk 4]
```diff
@@ -199,24 +201,30 @@ impl Contains<Location> for LocalPlurality {
...
 /// The barriers one of which must be passed for an XCM message to be executed.
-pub type Barrier = TrailingSetTopicAsId<(
-    // Weight that is paid for may be consumed.
-    TakeWeightCredit,
-    // Expected responses are OK.
-    AllowKnownQueryResponses<XcmPallet>,
-    WithComputedOrigin<
+pub type Barrier = TrailingSetTopicAsId<
+    DenyThenTry<
+        DenyReserveTransferToRelayChain,
```
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/orchestrator-relays/runtime/dancelight/src/xcm_config.rs:206]: Makes sense, will check if we can revert it.

File: chains/container-chains/runtime-templates/simple/src/lib.rs [hunk 5]
```diff
@@ -201,6 +205,22 @@ impl WeightToFeePolynomial for WeightToFee {
...
+        /// Network and location for the Ethereum chain. On Starlight, the Ethereum chain bridged
+        /// to is the Ethereum mainnet, with chain ID 1.
+        /// <https://chainlist.org/chain/1>
+        /// <https://ethereum.org/en/developers/docs/apis/json-rpc/#net_version>
+        pub EthereumNetwork: NetworkId = NetworkId::Ethereum { chain_id: 11155111 };
+        pub EthereumLocation: Location = Location::new(1, EthereumNetwork::get());
+}
+
+parameter_types! {
+    pub EthereumSovereignAccount: AccountId =
```
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/container-chains/runtime-templates/simple/src/lib.rs]: Indeed we don't need it.

File: chains/container-chains/runtime-templates/frontier/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/container-chains/runtime-templates/frontier/src/lib.rs:317]: @girazoki, any advice here?

[the remark this answers is not in the document]
Comment (girazoki) [reply chains/container-chains/runtime-templates/frontier/src/lib.rs:317]: work for the future. Agustin's idea seems the best as fetching it from the relay means making a lot of assumptions (the relay knows the ethereum network for insntace) which might not hold. parameters seems to be the appropriate one BUT it would need to be initialized in some sort, I would not make it part of this PR.

File: primitives/bridge/src/lib.rs [hunk 6]
```diff
@@ -374,6 +377,9 @@ where
             (1, [GlobalConsensus(Ethereum { chain_id })]) => {
                 Some(Self::from_chain_id(chain_id).into())
             }
+            (2, [GlobalConsensus(Ethereum { chain_id })]) => {
```
Comment (girazoki) [inline primitives/bridge/src/lib.rs:380]: while I do understand that you want to reduce the code, this should not belong here. it should be a separate struct. it is a bad idea to join the ethereum converter for a container chain with the assumptions of the solo-chain orchestrator.

File: primitives/bridge/src/container_token_to_ethereum_message_exporter.rs [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+        log::info!("validate params: network={network:?}, _channel={_channel:?}, universal_source={universal_source:?}, destination={destination:?}, message={message:?}, ");
+
+        if network != expected_network {
+            log::trace!(target: "xcm::ethereum_blob_exporter", "skipped due to unmatched bridge network {network:?}.");
+            return Err(NotApplicable);
+        }
+
+        // Cloning destination to avoid modifying the value so subsequent exporters can use it.
+        let dest = destination.clone().take().ok_or(MissingArgument)?;
+        if dest != Here {
```
Comment (girazoki) [inline primitives/bridge/src/container_token_to_ethereum_message_exporter.rs:94]: we should pass the para-id to the converter and dissallow anything that is sending a token that does not belong to that para

File: primitives/bridge/src/barriers.rs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            [Instruction::SetFeesMode { jit_withdraw: true }, Instruction::ExportMessage { .. }] => {
+                true
+            }
+            _ => false,
+        };
+
+        // Check all ExportMessage instructions
+        let mut all_exports_valid = false;
+        for instr in instructions {
+            if let Instruction::ExportMessage { xcm, .. } = instr {
```
Comment (girazoki) [inline primitives/bridge/src/barriers.rs:54]: I am starting to think this barrier is non-needed @Agusrodri. the message that we receive would look something like this:

- WithdawAsset
- BuyExecution
- ExportMessage

which means that it will pass the first barrier (AllowTopLevelPaidExecutionFrom). so we sould ask ourselves what do we want do here?

- A) we want to prevent a user from directly sending the exportMessage without transferring the tokens in the container-chain to the sovereign account. this should already be prevented because the DescendOrigin instruction would be the first instruction AFAIK, but we need to check that. maybe it's not prohibited and we need to do something smarter. But I think this should already be prevented as otherwise I could send to any other chain a ReserverAssetDeposited with an arbitrary amount deposited.

B) we want that only when tokens are transferred to the sovereign account in the container. if what I said above is true, one thing we can do is require always in the exporter that the origin calling the export is a parachain. I think this is the safest. in that case we DONT need this barrier whatsoever. 

Let me know what you think

Comment (girazoki) [reply primitives/bridge/src/barriers.rs:54]: my current line of thought is that Aauser should never be able to call exportMessage. and the only way to prevent that is not in the barrier, but in the exporter itself, by making sure universal_source belongs to a parachain

Comment (girazoki) [reply primitives/bridge/src/barriers.rs:54]: I think this line in the exporter makes sure of it but we should check it

```
let _ = match local_sub.as_slice() {
            [Parachain(para_id)] => *para_id,
            _ => {
                log::error!(target: "xcm::ethereum_blob_exporter", "could not get parachain id from universal source '{local_sub:?}'.");
                return Err(NotApplicable);
            }
        };
```
but we shoudl test it. basically the tests should cover:

- a user trying to call export locally in dancelight (pallet-xcm->execute->ExportMessage)
- a user trying to send a export message to dancelight from a container (pallet-xcm->send->withdrawAsset+BuyExecution+exportMessage)

File: chains/container-chains/runtime-templates/frontier/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/container-chains/runtime-templates/frontier/src/lib.rs:317]: Keep this hardcoded for now and added the [ticket](https://opslayer.atlassian.net/browse/MD-1277) for further improvement.

File: chains/xcm-integration-tests/emulated/tests/assets/dancebox/src/tests/mod.rs [hunk 9]
```diff
@@ -14,6 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Tanssi.  If not, see <http://www.gnu.org/licenses/>.
 
+mod container_chain_to_eth_transfer;
```
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/xcm-integration-tests/emulated/tests/assets/dancebox/src/tests/mod.rs]: Yes, makes sense. Fixed: https://github.com/moondance-labs/tanssi/pull/1100/commits/a7606be7dca9af0e1e9a131c368422172bcc2c74

File: chains/orchestrator-relays/runtime/dancelight/src/xcm_config.rs [hunk 10]
```diff
@@ -204,6 +206,7 @@ pub type Barrier = TrailingSetTopicAsId<(
     TakeWeightCredit,
     // Expected responses are OK.
     AllowKnownQueryResponses<XcmPallet>,
+    barriers::AllowExportMessageFromContainerChainBarrier,
```
[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/orchestrator-relays/runtime/dancelight/src/xcm_config.rs:209]: https://github.com/moondance-labs/tanssi/pull/1100/commits/80cab3ed9151efb556b4a2cc910737603492faac

File: primitives/bridge/src/barriers.rs (hunk 8 above)
Comment (evgeny-s) [reply primitives/bridge/src/barriers.rs]: sure, will update it.

Comment (Agusrodri) [reply primitives/bridge/src/barriers.rs]: Cool! Looks much better now.

File: chains/container-chains/runtime-templates/simple/src/xcm_config.rs [hunk 11]
```diff
@@ -84,6 +88,9 @@ parameter_types! {
     pub const BaseDeliveryFee: u128 = 100 * MICROUNIT;
 
     pub RootLocation: Location = Location::here();
+
+    // TODO: Revisit later
+    pub const ContainerToEthTransferFee: u128 = 2_700_000_000_000u128;
```
[the remark this answers is not in the document]
Comment (girazoki) [reply chains/container-chains/runtime-templates/simple/src/xcm_config.rs:93]: is it what we charge in tanssi for relaying?

[the remark this answers is not in the document]
Comment (evgeny-s) [reply chains/container-chains/runtime-templates/simple/src/xcm_config.rs:93]: So, the transfer fee can't be zero. It has some minimum value that we should respect.

File: primitives/bridge/src/container_token_to_ethereum_message_exporter.rs [hunk 12]
```diff
@@ -0,0 +1,1679 @@
...
+    snowbridge_core::{AgentId, ChannelId, TokenId},
+    snowbridge_outbound_queue_primitives::v1::message::{Command, Message, SendMessage},
+    sp_core::H160,
+    sp_runtime::traits::MaybeEquivalence,
+    xcm::latest::SendError::{MissingArgument, NotApplicable, Unroutable},
+    xcm::prelude::*,
+    xcm_executor::traits::ExportXcm,
+};
+
+pub struct EthereumBlobExporter<
```
Comment (girazoki) [inline primitives/bridge/src/container_token_to_ethereum_message_exporter.rs:31]: I would change the name of this directly to ContainerEthereumBlobExporter

Comment (evgeny-s) [reply primitives/bridge/src/container_token_to_ethereum_message_exporter.rs]: Updated here: https://github.com/moondance-labs/tanssi/pull/1100/commits/fd2bed662b7a6e5eacdb2f9572635fca8e08e337

**Comment by evgeny-s** [comment]
> the token-id conversion (from location to token-id) should be changed to gather it from ethereumSystem. this should:
prevent us from sending a token that is not registered
prevent us from failures if a new xcm version changes locations, as we only need to migrate ethereumSystem.

Fixed here: https://github.com/moondance-labs/tanssi/pull/1100/commits/314f337a009c95bead47473a1b1b9e5738cce8e6

File: primitives/container-chain/src/sovereign_paid_remote_exporter.rs [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        // `xcm` should already end with `SetTopic` - if it does, then extract and derive into
+        // an onward topic ID.
+        let maybe_forward_id = match xcm.last() {
+            Some(SetTopic(t)) => Some(*t),
+            _ => None,
+        };
+
+        let export_instruction = ExportMessage {
+            network: EthereumNetwork::get(),
+            destination: Here,
```
Comment (girazoki) [inline primitives/container-chain/src/sovereign_paid_remote_exporter.rs:72]: isnt the destination here wrong? should it be the ethereum chain?
